Implementing auto creation of streams

JacobASeverson opened the pull request:
This pull request adds ability for producers and consumers to create a stream if it isn't already existing. I've used the property `spring.cloud.stream.kinesis.binder.autoCreateStreams` to trigger the process. I've intentionally kept the scope to auto-creation only and not included shard manipulation in this PR. Included are passing unit tests and an integration test that works if a local kinesalite instance is running. I also have a example project at https://github.com/JacobASeverson/stream-example-kinesis that I've used to verify the functionality within AWS.

* Added ability to auto-create streams to the provisioner

* Added unit testing for the provisioner new stream handler

* Added an integration test for binding

Resolves spring-cloud/spring-cloud-stream-binder-aws-kinesis#6

artembilan commented on spring-cloud-stream-binder-kinesis/src/main/java/org/springframework/cloud/stream/binder/kinesis/config/KinesisBinderConfiguration.java:
```diff
@@ -59,8 +60,13 @@ public AmazonKinesisAsync amazonKinesis(
-	public KinesisStreamProvisioner provisioningProvider(AmazonKinesisAsync amazonKinesis) {
-		return new KinesisStreamProvisioner(amazonKinesis, this.configurationProperties);
+	public KinesisStreamHandler kinesisStreamHandler(AmazonKinesisAsync amazonKinesis) {
+		return new KinesisStreamHandler(amazonKinesis);
```
I wouldn't introduce extra level if we can just stick with the existing clear API from the `AmazonKinesis`.

artembilan commented on spring-cloud-stream-binder-kinesis/src/test/java/org/springframework/cloud/stream/binder/kinesis/KinesisBinderTests.java:
```diff
@@ -18,13 +18,15 @@
-import org.springframework.cloud.stream.binder.PartitionCapableBinderTests;
-import org.springframework.cloud.stream.binder.Spy;
+import org.junit.Test;
+import org.springframework.cloud.stream.binder.*;
```
No asterisk imports, please.

artembilan commented on spring-cloud-stream-binder-kinesis/src/test/java/org/springframework/cloud/stream/binder/kinesis/LocalKinesisResource.java:
```diff
@@ -53,7 +49,6 @@ protected void obtainResource() throws Exception {
 		System.setProperty(SDKGlobalConfiguration.AWS_CBOR_DISABLE_SYSTEM_PROPERTY, "true");
 
 		this.resource = AmazonKinesisAsyncClientBuilder.standard()
-				.withCredentials(mock(AWSCredentialsProvider.class))
```
Why did you removed this?
How does it work then?

I didn't check how the Kinesalite works because I'm on Windows 😄 , but I don't believe that we can stay with something default for credentials.

JacobASeverson replied:
Giving the client a stock proxy object doesn't provide any credentials despite the client expecting them. I was getting credential errors until I removed that line. 

I didn't dig deeper into the authentication mechanics of Kinesalite, but I confirmed the interactions were working without specifying credentials.

artembilan replied:
I don't understand how does this work for you.
I have:
```
org.junit.AssumptionViolatedException: Skipping test due to KINESIS not being available com.amazonaws.SdkClientException: Unable to load AWS credentials from any provider in the chain
```
and it is still too far from the real connection to the `kinesalite`.

When I make it like this:
```
.withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials("", "")))
```
The connection to my local `kinesalite` in the Nod.js starts to work.
And voila!, I have some tests passed, but most of them don't work anyway. 😢

JacobASeverson replied:
I'm not sure why removing credentials doesn't work in your environment, but it seems using the `AWSStaticCredentialsProvider` also works for me - I'll add it since it seems to be a common denominator.

artembilan commented on spring-cloud-stream-binder-kinesis-core/src/main/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisioner.java:
```diff
@@ -40,34 +40,40 @@
+		logger.info("Using Kinesis stream for outbound: " + name);
 
-		KinesisProducerDestination producer = new KinesisProducerDestination(name);
+		KinesisStream stream = kinesisStreamHandler.createOrUpdate(name, 1,
```
The `KinesisExtendedBindingProperties` must be injected into the `KinesisMessageChannelBinder` from the `KinesisBinderConfiguration` to pull the properties provided by end-user.

sobychacko replied:
As @artembilan suggested, instead of hard-coding the value `1` for the shard counts, get it from the provided producer and consumer properties.

artembilan commented on spring-cloud-stream-binder-kinesis/src/test/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisionerTests.java:
```diff
@@ -0,0 +1,117 @@
+/**
+ * @author Jacob Severson
+ */
+public class KinesisStreamProvisionerTests {
```
IMO this test class must be in the `spring-cloud-stream-binder-kinesis-core` module - close to the `KinesisStreamProvisioner`

JacobASeverson commented on spring-cloud-stream-binder-kinesis-core/src/main/java/org/springframework/cloud/stream/binder/kinesis/properties/KinesisBinderConfigurationProperties.java:
```diff
@@ -26,4 +26,13 @@
+	private boolean autoCreateStreams = true;
+
+	public boolean getAutoCreateStreams() {
+		return autoCreateStreams;
```
Hi @artembilan , Thanks for taking the time to review! This implementation is heavily influenced by the Kafka binder - there are a handful of things I questioned myself, but when in doubt I tried to stay consistent with that project.

`autoCreateStreams` is modeled after the Kafka binder's `autoCreateTopics` property [detailed here](http://docs.spring.io/autorepo/docs/spring-cloud-stream-binder-kafka-docs/1.1.0.M1/reference/htmlsingle/#_kafka_binder_properties). I also don't see much value in allowing the binder to fail when there's no existing stream and `autoCreateStreams`=`false`, but it is consistent with the [Kafka binder's handling](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/blob/master/spring-cloud-stream-binder-kafka-core/src/main/java/org/springframework/cloud/stream/binder/kafka/provisioning/KafkaTopicProvisioner.java#L118). 

Let me know if it's preferable to refactor that property out and default the provisioner to create/update on startup - I'd be all for it.

artembilan replied:
Great, @JacobASeverson ! 

I'm glad that we with you are in the same boat.

I wasn't involved in the Kafka Binder design, so can't say what is the point to allow to have that behavior there. Maybe main point that Kafka allows to create topic on demand by its nature: http://kafka.apache.org/documentation.html#configuration. See `auto.create.topics.enable`. Therefore even if we configure there `autoCreateStreams=false` and no topic on the Broker in advance, it will be created automatically by the first interaction with it - consume or produce.

Here in Kinesis we don't have such a luxury, hence the confuse with this option.

So, we should create stream in the Kinesis if it doesn't exist unconditionally.
Although I guess we could have `updateShardCount = true/false` option and call an appropriate Kinesis method if provides shards count isn't equal to the existing in stream. Otherwise we just return the stream as is overriding provided `partitions` options.

JacobASeverson commented:
Hi @artembilan, I believe I've addressed the first round of comments. Thanks!

JacobASeverson commented on spring-cloud-stream-binder-kinesis-core/src/main/java/org/springframework/cloud/stream/binder/kinesis/provisioning/KinesisStreamProvisioner.java:
```diff
@@ -55,19 +61,40 @@ public KinesisStreamProvisioner(AmazonKinesis amazonKinesis,
+			logger.info("Stream found, using existing stream");
 
-		KinesisConsumerDestination consumer = new KinesisConsumerDestination(name);
+			return new KinesisStream(name, streamResult.getStreamDescription().getShards().size());
```
Right, `createOrUpdate` was intended to house shard creation/manipulation eventually, but I intentionally wanted to limit the scope of this work to stream auto-creation. This is also why I didn't fix the integration tests dealing with Kinesalite that didn't seem to deal with this end-goal. I agree with your idea of an `updateShardCount` property, but similarly left it out to be part of another effort.

If you would like such shard handling and finishing of integration tests to be part of this effort let me know and I'll look at those as well.

artembilan replied:
Do not loose the track of the concern I've just raised https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/issues/16

If you aware about all those problems with the integration tests, let's raise an addition GH issue issue as well!

Then right, I will concentrate on what you've done so far and that's all.

Sorry for misunderstanding, but your contribution style is a bit unusual 😄 
But thank you anyway!

JacobASeverson replied:
No problem, I'm taking a narrow approach for this contribution because I'm just learning both Spring Cloud Stream and Kinesis and don't want the PR to be open for too long. Thank you for your patience! :-)

artembilan commented on spring-cloud-stream-binder-kinesis/src/test/java/org/springframework/cloud/stream/binder/kinesis/KinesisBinderTests.java:
```diff
@@ -38,6 +45,21 @@
+		ExtendedConsumerProperties<KinesisConsumerProperties> consumerProperties = createConsumerProperties();
+		String testTopicName = "nonexisting" + System.currentTimeMillis();
+		Binding<?> binding = binder.bindConsumer(testTopicName, "test", output, consumerProperties);
+		binding.unbind();
```
This has not been addressed.
I'm fine with the rest of changes, but would like to see some assertion here anyway.
E.g. assert that `binding` has right properties.

JacobASeverson commented:
Hi @artembilan, I've updated the binder test to assert stream creation and also updated my branch with your new checkstyle rules. 

One thing I want to highlight is that I've added a `connectTimeout` property to `KinesisConsumerProperties` to allow the user to specify how long to retry connection to Kinesis before failing on startup. Since stream creation can take some time before being `ACTIVE`, I was seeing the consumer creating the stream but then eating a `ResourceNotFound` exception which meant the consumer application would start up successfully but not consume any messages. Using the timeout in this manor will ensure that the consumer is either able create the stream and consume messages or it won't start successfully.

artembilan commented on spring-cloud-stream-binder-kinesis/src/main/java/org/springframework/cloud/stream/binder/kinesis/KinesisMessageChannelBinder.java:
```diff
@@ -108,8 +108,9 @@ protected MessageProducer createConsumerEndpoint(ConsumerDestination destination
-		adapter.setDescribeStreamRetries(1);
+		adapter.setStartTimeout(properties.getExtension().getConnectTimeout());
+		// Let the start timeout interrupt startup instead of eating the ResourceNotFoundException
+		adapter.setDescribeStreamRetries(properties.getExtension().getConnectTimeout() + 10);
```
That's something what isn't good: `10000 + 10` retries!
Maybe let end-user to configure this option for `KinesisMessageDrivenChannelAdapter` as well?
I'd really prefer to expose all of them, but we definitely can address that in the separate issue, e.g. #9 

But for now to avoid confuses and magic numbers let's expose `describeStreamRetries` as well!
OTOH I called the property name as `startTimeout` in the `KinesisMessageDrivenChannelAdapter`, not sure why should we rename it here into `connectTimeout` since it is really not about connection...

I don't fail channel adapter for non-ACTIVE stream, because it can be configured for several of them and others might be active and fully available.

Hope this makes sense to you.

Thanks

artembilan commented on spring-cloud-stream-binder-kinesis/src/test/java/org/springframework/cloud/stream/binder/kinesis/KinesisBinderTests.java:
```diff
@@ -16,28 +16,58 @@
+
+		Assert.assertThat(createdStreamName, Is.is(testStreamName));
+		Assert.assertThat(createdShards, Is.is(Matchers.greaterThan(0)));
+		Assert.assertThat(createdStreamStatus, Is.is("ACTIVE"));
```
Static imports for JUnit, Hamcrest and Mockito are fine in tests:
```
<module name="AvoidStaticImport">
            <property name="excludes"
                      value="org.junit.Assert.*,org.mockito.Mockito.*,org.mockito.Matchers.*,org.hamcrest.Matchers.*,org.assertj.core.api.Assertions.*"/>
</module>
```

More over we there particular suppression:
```
<suppress files="[\\/]src[\\/]test[\\/]java[\\/]" checks="AvoidStaticImport" />
```

So, consider do not refactor this concern in the future, please.

artembilan commented on spring-cloud-stream-binder-kinesis-core/pom.xml:
```diff
@@ -26,6 +26,11 @@
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-stream-binder-test</artifactId>
```
Any particular reason to use this dependency?
I don't see any explanation in the commit messages, too.

Thanks

artembilan commented on spring-cloud-stream-binder-kinesis-core/src/main/java/org/springframework/cloud/stream/binder/kinesis/properties/KinesisConsumerProperties.java:
```diff
@@ -19,8 +19,18 @@
+	private int connectTimeout = 60000;
+
+	public int getConnectTimeout() {
+		return connectTimeout;
```
I know that `requireThis` rule isn't enabled in the SCSt's `checkstyle.xml`, but would be great to keep it on board.
So, the code here must be:
```
return this.connectTimeout;
```
At least it is much easier to review from here on GH.

artembilan commented on spring-cloud-stream-binder-kinesis/src/test/java/org/springframework/cloud/stream/binder/kinesis/KinesisBinderTests.java:
```diff
@@ -16,28 +16,58 @@
+		String createdStreamStatus = streamResult.getStreamDescription().getStreamStatus();
+
+		Assert.assertThat(createdStreamName, Is.is(testStreamName));
+		Assert.assertThat(createdShards, Is.is(Matchers.greaterThan(0)));
```
Can't we assert against the real number of shards we have requested?

artembilan commented:
Merged as https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/commit/d4192aea915e326317e58e1200f4673f6cdc2d49.

@JacobASeverson ,
thank you very much for the contribution.
Looking forward for more!

You can choose any open issue for work or just PR anything you think is important or missed.

That's great to a progress!

JacobASeverson commented:
@artembilan thanks for your help!

JacobASeverson commented:
@artembilan minor question - is it possible to associate my username to those commits? It's not a big deal, but nice to have contributions on record.

artembilan commented:
Yes, of course! That's fully valid request.

I tried to reinstate your authority on the commit and now it is like this:
>  Jacob Severson committed with artembilan an hour ago

Looks like I made some typo there, sorry. Will be more careful in the future!

JacobASeverson commented:
It looks like my email address had a typo, if it's something you'd be willing to do, I think you can fix the author portion by doing

`git commit --amend --author="Jacob Severson <jacoba.severson@gmail.com>"` and pushing?

artembilan commented:
Fixed.

Thank you for the hint!

JacobASeverson commented:
Looks good, thanks again! 😄
